bot/exts/utilities/bookmark.py

```diff
@@ -25,6 +23,49 @@
 )
 
 
+class LinkTargetMessage(discord.ui.View):
+    """The button that relays the user to the bookmarked message."""
+
+    def __init__(self, target_message: discord.Message):
+        super().__init__()
+        self.add_item(discord.ui.Button(label="View Message", url=target_message.jump_url))
+
+
+class SendBookmark(discord.ui.View):
+    """The button that sends the bookmark to other users."""
+
+    def __init__(
+        self,
+        action_bookmark_function: Callable[[discord.TextChannel, discord.Member, discord.Message, str], None],
+        author: discord.Member,
+        channel: discord.TextChannel,
+        target_message: discord.Message,
+        title: str
+    ):
+        super().__init__()
+
+        self.bookmark_function = action_bookmark_function
+        self.clicked = [author.id]
+        self.channel = channel
+        self.target_message = target_message
+        self.title = title
+
+    @discord.ui.button(label="Receive Bookmark", style=discord.ButtonStyle.green)
+    async def button_callback(self, button: discord.Button, interaction: discord.Interaction) -> None:
+        """The button callback."""
+        if interaction.user.id in self.clicked:
+            await interaction.response.send_message(
+                "You have already received a bookmark to that message.",
+                ephemeral=True,
+            )
+            return
+
+        await self.bookmark_function(self.channel, interaction.user, self.target_message, self.title)
+
+        await interaction.response.send_message("You have received a bookmark to that message.", ephemeral=True)
+        self.clicked.append(interaction.user.id)
```

Comment:
There's a possible race condition here if a user clicks the button multiple times in very short succession (assuming that's possible with ratelimits). In that case, we send out responses multiple times before the `self.clicked.append` line is hit. The fix would probably be appending to `clicked` directly below the `if` statement block.
Since we expect uniqueness of the users, `self.clicked` should also probably be a set -> `self.clicked = {author.id}`.

Reply:
When you click a button, Discord doesn't allow you to click it again until the app responds or times out

Reply:
This assumes that Discord's side of implementation is bug and race condition-free.

As far as I recall, the Discord API documentation mentions somewhere that events may be delivered multiple times as well. I cannot find it though.

Reply:
How should I go about this?

Reply:
Resolved in 273faec